From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: John Doe <john.doe@somewhere.on.planet>
Date: Sat, 27 Apr 2024 14:54:18 +0000
Subject: [PATCH] Allwinner: Fix incorrect ram size detection for T507

Signed-off-by: John Doe <john.doe@somewhere.on.planet>
---
 arch/arm/mach-sunxi/dram_sun50i_h616.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/arch/arm/mach-sunxi/dram_sun50i_h616.c b/arch/arm/mach-sunxi/dram_sun50i_h616.c
index d2e26993ea..b8fa9ff08d 100644
--- a/arch/arm/mach-sunxi/dram_sun50i_h616.c
+++ b/arch/arm/mach-sunxi/dram_sun50i_h616.c
@@ -1365,10 +1365,12 @@ static void mctl_auto_detect_dram_size(const struct dram_para *para,
 	/* detect row address bits */
 	config->cols = 8;
 	config->rows = 18;
 	mctl_core_init(para, config);
 
+    udelay(50);
+
 	for (config->rows = 13; config->rows < 18; config->rows++) {
 		/* 8 banks, 8 bit per byte and 16/32 bit width */
 		if (mctl_mem_matches((1 << (config->rows + config->cols +
 					    4 + config->bus_full_width))))
 			break;
@@ -1376,10 +1378,12 @@ static void mctl_auto_detect_dram_size(const struct dram_para *para,
 
 	/* detect column address bits */
 	config->cols = 11;
 	mctl_core_init(para, config);
 
+    udelay(50);
+  
 	for (config->cols = 8; config->cols < 11; config->cols++) {
 		/* 8 bits per byte and 16/32 bit width */
 		if (mctl_mem_matches(1 << (config->cols + 1 +
 					   config->bus_full_width)))
 			break;
-- 
Created with Armbian build tools https://github.com/armbian/build

